source/util.py

```diff
@@ -12,6 +12,77 @@
 # credibility in binary classification is (p-of-most-probable-class)
 
 
+
+def run_experiments(cp_in, train, test, calib, eps=0.1, times=1, rseed=None, method=''):
+    from copy import deepcopy
+    
+    if rseed is not None:
+        np.random.seed(rseed)
+    preds_dfs = []
+    for run in range(times):
+        cc = deepcopy(cp_in)
+        cc.fit(train, calib)
+        preds = []
+        verds = []
+        for ex in test:
+            pred = cc.predict(ex, eps)
+            verds.append(
+                pred.verdict(ex.get_class().value, 
+                             eps))
+            preds.append(pred)
+        preds_df = preds_to_df(preds)
+        preds_df['verdict'] = verds
+        add_inst_to_pred_df(preds_df, 
+                            test)
+        preds_dfs.append(
+            df_pred_inst_metrics(preds_df, 
+                                 dataset_name=test.name,
+                                 mondrian=cc.mondrian,
+                                 method=method, 
+                                 model_for_setup=cc))
+    return pd.concat(preds_dfs, axis=1).T
+
+
+
+
+def plot_experiments(df_experiments, data_title=None, scale=True):
+    import matplotlib.pyplot as plt
+    
+    is_mondrian = (lambda x: ' (Mondrian)' if x else '')
+    is_scaled = (lambda x: '\n(Confidence/Credibility Scaled)' if x else '')
+    fig, ax = plt.subplots()
+
+    for i, run in df_experiments.iterrows():    
+        plot_conf_cred(run.df, 
+                       title=f'{run.classifier}{is_mondrian(run.mondrian)}\ndata: {run.data if data_title is None else data_title}{is_scaled(scale)}', 
+                       ax=ax,
+                       scale=scale,
+                       alpha=0.5)
+        ax.hlines(y=run.verdict, 
+                  xmin=0, 
+                  xmax=len(run.df), 
+                  label=f'acc {run.verdict*100:.2f}%', 
+                  color='g', 
+                  linestyles='--')
+        ax.vlines(x=len(run.df)/4, ymin=0, ymax=1, color='k', lw=3, linestyles=(0, (1, 10)))
+        ax.vlines(x=len(run.df)/2, ymin=0, ymax=1, color='k', lw=3, linestyles=(0, (1, 10)))
+        ax.vlines(x=len(run.df)*3/4, ymin=0, ymax=1, color='k', lw=3, linestyles=(0, (1, 10)))
+        ax.legend()
+
+
+
+
+def get_low_confidence_predictions(df, min_conf=0.95):
```

Comment:
We talked about returning a percent of the low confidence predictions. Maybe something like this could be another function we use is something like this.

The only issue I can think of with using the lowest percent is it will arbitrarily pick which ones to return if the conficence is the same across the percent we pick.

for example. Say we have samples 1-5 that have a confidences of exactly .7  and samples 6-10  had a confidence > .7. If we gave this function .2 for the lowest_percent, I think it would essentially randomly return any 2 samples from 1-5. I can't really think of a good work around for this. I we will just need to weigh what is more important to us. Do we want the same number returned each time, or do we want more consistency in the confidence that we return.

`def get_low_confidence_predictions(df, lowest_percent=0.1):
	num_to_return = math.ceil(len(df) * lowest_percent)
	#You'd have to import math, 
	#Otherwise just do this
	#num_to_return = round(len(df) * lowest_percent)
	df = df.sort_values(by=['confidence'])
	return df.head(num_to_return)`

Reply:
I agree and you make a very solid case here.

I do not see a concise, foolproof way to resolve the issue that arises when two or more predictions have the same value(s) and straddle the border of the specified threshold. However, I do see a consistent way to handle the situation. If we docstring the expected behavior, I would say we've done our due diligence.

Let me know what you think of below.

def get_low_confidence_predictions(df, percentile=10):
#
#  check valid range
assert 0 < percentile and percentile < 100, 'percentile outside valid range'
#
#  integer math rounds down, preventing rounding up beyond bounds for small collections (though .head() handles)
#  open to reasons to round up
num_to_return = len(df) // percent
#
#  when predictions with the same confidence value span the percentile boundary, 
#  credibility will be used as the secondary sort criteria
#
#  when predictions with the same confidence and credibility span the percentile boundary, 
#  predictions will be returned based upon their index values in the provided DataFrame
df = df.sort_values(by=['confidence', 'credibility'], kind='mergesort')
return df.head(num_to_return)

Reply:
I think that is a great solution. Let's do that. So do you think we will essentially replace the old way of selecting for lowest confidence with this?

Reply:
I found it consistent and really straight forward. I use it in `util.plot_race_representation_from_experiment`. I will actually commit a notebook really quick that is bare-bones experiment results--visualized

Reply:
If you see anywhere in particular where this makes sense to use, add the usage, run the code to verify, and feel free to commit